### ISSUE TITLE
android 11 (targetSdkVersion30) cant open sqlite database

### ISSUE THREAD

**Issue by farfromrefug** [CartoDB/mobile-sdk#469]
With the new scope storage permission system in android 11, carto mobile SDK cant open sqlite databases even if the app has the permissions for it.
All libraries are facing the issue. Here is a example explaining it a bit:
https://github.com/OutSystems/Cordova-sqlite-storage/pull/7

They have fixed it here https://github.com/OutSystems/Android-sqlite-native-driver/pull/1/files

**Comment by farfromrefug** [comment]
@mtehver yes i did. Simply try to open a mbtiles from the sdcard (not in android app folder) or outside of any folder of your app.
It is not related to the phone but to scope storage https://www.androidcentral.com/what-scoped-storage
Using the SDK i see that `sqlite3_open_v2` fails saying it cant open the database.
The linked cordova issue is the only place where i could find reference to the issue with `sqlite3_open_v2`. They mention it works with `SQliteAndroidDatabase` but we cant use this.
TBH i dont 100% understand how they solved it with https://github.com/OutSystems/Android-sqlite-native-driver/pull/1/files but i think the fix is there

**Comment by mtehver** [comment]
@farfromrefug Looking at the referenced code I do not understand the fix either, they seem to use Android built-in sqlite as a workaround. Why/how this works is not clear to me.

IMHO, the only sensible way is to simply comply to Android requirements here (and store everything under the directory Android provides via `getExternalFilesDir`). For other cases the workaround is to use target SDK 29 or try to hack the access via the Scoped Storage APIs.

**Comment by farfromrefug** [comment]
@mtehver No they first used the `oldImpl` of using `SQliteAndroidDatabase` but then they changed they native sqlite implementation to allow opening file when they have the permission.
Here is the PR when they rolledback and correctly implemented it https://github.com/CartoDB/mobile-sdk/issues/469 They simply used their new  native implementation which apparently fixes it (Which is why i linked to it).

There must be a way to make it work. [Here](https://github.com/storesafe/cordova-sqlite-storage/issues/954#issuecomment-716970227) they talk about a return value having changed. Are we using the platform sqlite lib? Might they have update the platform sqlite lib for it to work with scope storage?

**Comment by mtehver** [comment]
@farfromrefug You mean the fix here: https://github.com/OutSystems/Android-sqlite-native-driver/pull/1/files ?
This simply fixes a hack related to a JNI return type wrapping, nothing more. The discussion threads are really difficult to follow, there is so much noise. It is not clear where the database files are stored, for example.

CARTO SDK uses its own copy of Sqlite, Android does not allow linking to the built-in system Sqlite library.

The main question I have is - if you use `getExternalFilesDir` to get a directory where to store .mbtiles files or offline packages, does everything work with target SDK 30 the same way as with earlier target SDK versions? If that is the case, I do not see anything that needs to be fixed.

**Comment by farfromrefug** [comment]
@mtehver agree i linked the PR i case you would understand it :D
The issue is i "cant" copy my mbtiles to a dir from my app. It is around 10gb!
I need to be able to read mbtiles from sdcard directly just like we did before android 30.
And as `SQliteAndroidDatabase` seems to be able to do it there must be a way.
Remember that i try to access the sqlite database after having asked the permission and used [`takePersistableUriPermission`](https://developer.android.com/reference/android/content/ContentResolver#takePersistableUriPermission(android.net.Uri,%20int))
I will dig further

**Comment by binu8051** [comment]
@farfromrefug do you got any solution?
Even i tried to placed the Sqlite files on Documents folder.

**Comment by farfromrefug** [comment]
@binu8051 sadly not yet. was thinking on searching in mapbox repo to see of they found a solution. they use sqlite too

**Comment by farfromrefug** [comment]
@mtehver i made some research and looking at android doc they mention [here](https://developer.android.com/about/versions/11/privacy/storage#media-direct-file-native) that `fopen` should work. Wondering if we could use it with  `sqlite3_open_v2` instead of giving the path directly.
What do you think?

**Comment by mtehver** [comment]
@farfromrefug I currently do not have time to look into this myself. I doubt that converting internal of `sqlite3_open_v2` to use `fopen` would change anything, as `fopen` is simply a wrapper to normal `open` with buffering support.

A similar issue is described here: https://stackoverflow.com/questions/59000390/android-accessing-files-in-native-c-c-code-with-google-scoped-storage-api/59004193#59004193

If I understand correctly, then the following snippet *may* work after you have gained the access rights to the file:

```
String getFileNameThatICanUseInNativeCode(Uri uri) throws FileNotFoundException {
      mParcelFileDescriptor =
              getApplicationContext().getContentResolver().openFileDescriptor(uri, "r");
      if (mParcelFileDescriptor != null) {
        int fd = mParcelFileDescriptor.getFd();
        File file = new File("/proc/self/fd/" + fd);
        String path = null;
        try {
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            path = Os.readlink(file.getAbsolutePath()).toString();
          }
        } catch (ErrnoException e) {
          e.printStackTrace();
        }

        return path;
      }
      else{
        return null;
      }
  }
```

**Comment by farfromrefug** [comment]
@mtehver the method as it is would not work as it only return the "old way" path (`/storage/...` instead of `content://...`)
I tried to return `"/proc/self/fd/" + fd` instead and it still would not work. Though `openFileDescriptor` works and it seems manage to open the file for reading
Not sure where to go from there

**Comment by mtehver** [comment]
@farfromrefug If you really need this, then perhaps you could look into sqlite3.c and try to do manual parsing of input file name (extracting FD from '/proc/self/fd/FD') and then replace `open(filename)` with `fdopen(FD)` in the code.

**Comment by farfromrefug** [comment]
@mtehver thanks. that file is hard to read. Still trying to find the right `open()` to modify ;)

**Comment by krishna5151** [comment]
android 11 (targetSdkVersion 32) cant open sqlite database in ionic 3 .

 cordova plugins are installed. but we are get this error
'Ionic Native: tried accessing the SQLite plugin but it's not installed.
vendor.js:1 Install the SQLite plugin: 'ionic cordova plugin add cordova-sqlite-storage'

package.json : 

 "@ionic-native/sqlite": "^4.20.0",
"cordova-sqlite-storage": "^6.0.0",

database.ts :

    if (!this.isOpen) {
      this.storage = new SQLite();
      this.storage.create({ name: "data.db", location: "default" }).then((db: SQLiteObject) => {
        this.db = db;
        console.log('DB :',this.db)
        db.executeSql("CREATE TABLE IF NOT EXISTS customer (id_customer int UNSIGNED NOT NULL, reference_no varchar(100) DEFAULT NULL)", []);
        this.isOpen = true;
        console.log('open : ',this.isOpen)
      }).catch((error) => {
        console.log('error :',error);
      })
    }
